src/main/java/org/jabref/gui/citationkeypattern/GenerateCitationKeyAction.java

```diff
@@ -80,30 +83,51 @@ private void checkOverwriteKeysChosen() {
         }
     }
 
-    private void generateKeys() {
-        if (isCanceled) {
-            return;
-        }
-
-        stateManager.getActiveDatabase().ifPresent(databaseContext -> {
-            // generate the new citation keys for each entry
-            final NamedCompound compound = new NamedCompound(Localization.lang("Autogenerate citation keys"));
-            CitationKeyGenerator keyGenerator =
-                    new CitationKeyGenerator(databaseContext, Globals.prefs.getCitationKeyPatternPreferences());
-            for (BibEntry entry : entries) {
-                keyGenerator.generateAndSetKey(entry)
-                            .ifPresent(fieldChange -> compound.addEdit(new UndoableKeyChange(fieldChange)));
-            }
-            compound.end();
-
-            // register the undo event only if new citation keys were generated
-            if (compound.hasEdits()) {
-                frame.getUndoManager().addEdit(compound);
+    private BackgroundTask generateKeysInBackground() {
+        return new BackgroundTask<Void>() {
+
+            @Override
+            protected Void call() {
+
+                    if (isCanceled) {
+                        return null;
+                    }
+                    showToUser(true);
+                    titleProperty().set(Localization.lang("Autogenerate citation keys"));
+                    messageProperty().set(entries.size() + " " + Localization.lang("entries"));
+
+                    DefaultTaskExecutor.runInJavaFXThread(() -> {
```

Comment:
The runInJavaFXThread shouldn't be necessary as this is handled by updateProgress.

Reply:
I wondered about that, but it is done in ImportHandler as well. Shall I remove it there as well (in an additional PR)?

Reply:
For some reason it is not working in this context, see https://github.com/JabRef/jabref/pull/7209/files#r551271470. Maybe @Siedlerchr still remembers where the problem was.

Reply:
The relevant code is https://github.com/JabRef/jabref/blob/fa8cc829727ffddd97a9056fd466e7098577f4d7/src/main/java/org/jabref/gui/util/DefaultTaskExecutor.java#L146-L155, but I don't see any problem with this.

Reply:
The progress updating is not happening in the fx thread by the BackgroundTask. This will leead to an exception.

Reply:
So the problem is that the properties on the BackgroundTask are updating on the background thread? Why is this a problem? Who is consuming these properties?

Reply:
So shall we explicitly make updating the progress variable run on the fx thread in updateProgress?
`
protected void updateProgress(BackgroundProgress newProgress) {

    DefaultTaskExecutor.runInJavaFXThread(() -> {

        progress.setValue(newProgress);

    });

}
`

Reply:
Ideally yes it would happend in the BackgroundTask itself, but I am unsure if this has any side effects on existing tasks

Reply:
would nesting DefaultTaskExecutor.runInJavaFXThread() be problematic (in addition to being ugly)?
I don't think there are many tasks that use updateProgress right now, let me investigate.

Reply:
As far as I (and Intellij) can tell, updateProgress is only used by GenerateCitationKeyAction, FileDownloadTask and ImportHandler.
FileDownloadTask uses EasyBind to update the progress. Does EasyBind already take care of running on the JavaFX thread?
https://github.com/JabRef/jabref/blob/fa8cc829727ffddd97a9056fd466e7098577f4d7/src/main/java/org/jabref/gui/externalfiles/FileDownloadTask.java#L24-L39

Reply:
I'm not sure... then using `setValue` on `progress` still leads to the same problems (i.e. you cannot bind something to the progress property). Personally, I would leave the whole what happens on which thread hidden as long as possible, i.e. only the DefaultTaskExecutor cares about these things when converting a background task to a JavaFX task. 

I guess these problems we are discussing can be easily fixed by not using the progress/message properties of the background task anywhere.